packages/access-api/src/models/delegations.js

```diff
@@ -0,0 +1,137 @@
+import * as Ucanto from '@ucanto/interface'
+import {
+  delegationsToBytes,
+  bytesToDelegations,
+} from '@web3-storage/access/encoding'
+
+/**
+ * @typedef {import('@web3-storage/access/src/types').DelegationTable} DelegationRow
+ * @typedef {Omit<DelegationRow, 'inserted_at'|'updated_at'|'expires_at'>} DelegationRowUpdate
+ */
+
+/**
+ * @typedef Tables
+ * @property {DelegationRow} delegations
+ */
+
+/**
+ * @typedef {import("../types/database").Database<Tables>} DelegationsDatabase
+ */
+
+/**
+ * DelegationsStorage that persists using SQL.
+ * * should work with cloudflare D1
+ */
+export class DbDelegationsStorage {
+  /** @type {DelegationsDatabase} */
+  #db
+
+  /**
+   * @param {DelegationsDatabase} db
+   */
+  constructor(db) {
+    this.#db = db
+    // eslint-disable-next-line no-void
+    void (
+      /** @type {import('../types/delegations').DelegationsStorage} */ (this)
+    )
+  }
+
+  get length() {
+    return new Promise((resolve, reject) => {
+      this.#db
+        .selectFrom('delegations')
+        .select((e) => e.fn.count('cid').as('size'))
+        .executeTakeFirstOrThrow()
+        .then(({ size }) => {
+          if (typeof size === 'string') {
+            const sizeNumber = parseInt(size, 10)
+            if (isNaN(sizeNumber)) {
+              throw new TypeError(
+                `unable to determine size number of delegations table`
+              )
+            }
+            return sizeNumber
+          }
+          if (typeof size === 'bigint') {
+            if (size > Number.MAX_SAFE_INTEGER) {
+              throw new TypeError(`table size too big for js Number`)
+            }
+            return Number(size)
+          }
+          return size
+        })
+        .then(resolve)
+        .catch(reject)
+    })
```

Comment:
Many questions:

* Why avoiding `await`?
* Why not return a `BigInt`?

```suggestion
    const data = await this.#db
        .selectFrom('delegations')
        .select((e) => e.fn.count('cid').as('size'))
        .executeTakeFirstOrThrow()
    return BigInt(data.size)
```

Reply:
* couldn't do `await` when `length` was a property. but now its not so i'll adjust.
* will also make it a bigint native now that I'm making DelegationsStorage not be met by an `Array`

Reply:
made those changes in https://github.com/web3-storage/w3protocol/pull/427/commits/022dc2167647dcbabb35b6ab0450dafd9c708a84